drv/lpc55-update-server/src/main.rs

```diff
@@ -781,105 +795,91 @@ impl ServerImpl<'_> {
+                return Err(UpdateError::ImageMismatch);
+            }
+        } else {
+            // We can't get a length for stage0.
+            // Perhaps it has been corrupted by an earlier update attempt
+            // and we're desperately trying to fix it by writing a good stage0
+            // image.
+            // Ok to update if signature on stage0next is good.
             return Err(UpdateError::ImageMismatch);
-        }
+        };
         // compare flash page to cache
-        let cached =
-            self.fw_cache[0..flen / core::mem::size_of::<u32>()].as_bytes();
+        let cached = self.fw_cache[0..clen / SIZEOF_U32].as_bytes();
         let mut flash_page = [0u8; BYTES_PER_FLASH_PAGE];
-        for addr in (0..flen).step_by(BYTES_PER_FLASH_PAGE) {
-            let size = if addr + BYTES_PER_FLASH_PAGE > flen {
-                flen - addr
-            } else {
-                BYTES_PER_FLASH_PAGE
+        for addr in (0u32..(clen as u32)).step_by(BYTES_PER_FLASH_PAGE) {
+            // If we encounter an unreadable page in stage0 then
+            // ok to update if signature on stage0next later tests ok.
+            stage0
+                .read_bytes(addr, flash_page.as_bytes_mut())
+                .map_err(|_| UpdateError::ImageMismatch)?;
+            let Some(end) = (addr as usize).checked_add(BYTES_PER_FLASH_PAGE)
+            else {
+                return Err(UpdateError::ImageMismatch);
             };
-
-            indirect_flash_read(
-                &self.flash,
-                addr as u32,
-                &mut flash_page[..size],
-            )?;
-            if flash_page[0..size] != cached[addr..addr + size] {
+            if flash_page != cached[(addr as usize)..end] {
                 return Err(UpdateError::ImageMismatch);
             }
         }
+        // Images already match, don't wear out flash more than necessary.
         Ok(())
     }
 
-    // Looking at a region of flash, determine if there is a possible NXP
-    // image programmed. Return the length in bytes of the flash pages
-    // comprising the image including padding to fill to a page boundary.
-    fn flash_image_len(&self, span: &Range<u32>) -> Result<usize, UpdateError> {
-        let buf = &mut [0u32; 1];
-        indirect_flash_read(
-            &self.flash,
-            span.start + LENGTH_OFFSET as u32,
-            buf[..].as_bytes_mut(),
-        )?;
-        if let Some(len) = round_up_to_flash_page(buf[0]) {
-            // The minimum image size should be further constrained
-            // but this is enough bytes for an NXP header and not
-            // bigger than the flash slot.
-            if len as usize <= span.len() && len >= HEADER_OFFSET {
-                return Ok(len as usize);
+    fn cached_stage0_image_len(&self) -> Result<usize, UpdateError> {
+        if let Ok(vectors) = ImageVectorsLpc55::try_from(
+            self.fw_cache[0..BLOCK_SIZE_BYTES].as_bytes(),
```

Comment:
Slicing a `BLOCK_SIZE_BYTES` chunk then passing it into `read_from_prefix` doesn't make sense to me.

We should
1) Add a const assertion that `fw_cache` is long enough to fit `core::mem::size_of::<ImageVectorsLpc55>`
2) do something like 
```rust
let vectors = ImageVectorsLpc55::read_from(
    &self.fw_cache[..size_of::<ImageVectorsLpc55>()]
).unwrap_lite()
```
3) Remove `ImageVectorsLpc55`'s `TryFrom` implementation, since this is the only place it's used